[mlir][transform] Fix handling of transitive include in interpreter.

ingomueller-net opened the pull request:
This is a new attempt at fixing transitive includes in the transform dialect interpreter (next to #67241) and a preparation for being able to load multiple transform library files (#67120).

Until now, the interpreter would only load those symbols from the provided library files that were declared in the main transform module. However, sequences in the library may include other sequences on their own. Until now, if such sequences were not *also* declared in the main transform module, the interpreter would fail to resolve them. Forward declaring all of them is undesirable as it defeats the purpose of encapsulation into library modules.

This PR implements a kind of linker for transform scripts to solve this problem. The linker merges all symbols of the library module into the main module before interpreting the latter. Symbols whose names collide are handled as follows: (1) if they are both functions (in the sense of `FunctionOpInterface`) with compatible signatures, one is external, and the other one is public, then they are merged; (2) of one of them is private, that one is renamed; and (3) an error is raised otherwise.

TODO:
- [x] Enable running interpreter several times (see discussion below).
- [x] Fix several smaller issues marked with `XXX`.
- [x] Fix reproducer: since injection of the library is now not idempotent anymore, we should not print the filename argument in the repro call.
- [x] Revise CLI and function parameter documentation and description.

ingomueller-net committed:
[mlir][transform] Fix handling of transitive include in interpreter.

Until now, the interpreter would only load those symbols from the
provided library files that were declared in the main transform module.
However, sequences in the library may include other sequences on their
own. Until now, if such sequences were not *also* declared in the main
transform module, the interpreter would fail to resolve them. Forward
declaring all of them is undesirable as it defeats the purpose of
encapsulation into library modules.

This PR implements a kind of linker for transform scripts to solve this
problem. The linker merges all symbols of the library module into the
main module before interpreting the latter. Symbols whose names collide
are handled as follows: (1) if they are both functions (in the sense of
`FunctionOpInterface`) with compatible signatures, one is external, and
the other one is public, then they are merged; (2) of one of them is
private, that one is renamed; and (3) an error is raised otherwise.

ingomueller-net commented:
@ftynse: This is a new attempt at #67241 along the lines of your suggestions. Please take a look. It is worth mentioning that (i) I merged the stages (1) and (2) from your suggestion, and (ii) did not implement the optimization of only including needed symbols.

I have identified one more problem that may require a major change: it is currently not always possible to run the interpreter pass several times. The problem is the following: if the script contains a declaration `@foo` and the library contains a matching definition, then than definition is merge into the transform module upon its first invocation. In the second invocation, however, `@foo` is now a *definition* in the transform module, so merging in the library module leads to a `doubly defined symbol` error. This did not happen before because only the (directly) missing symbols were merged, and `@foo` is not missing after the first invocation anymore.

I see two ways out: (A) only merge in missing symbols (plus their dependencies) as previously and make all imported symbols private, or (B) separate the merging from the interpreter into a separate pass. I have the feeling that Approach (A) would make the merging again more special to the interpreter and possibly somewhat brittle and Approach (B) would be more generic and possibly more performant. (If the interpreter pass is really used several times, linking has to be done each time, while with a dedicated linking pass, it would only be done once.)

I imagine the linking pass from Approach (B) to be called something like `load-transform-dialect-interpreter-module` or similar, take the `transform-library-file-name` and `transform-library-file-name` arguments (as well as their extensions planned in #67120), and then only do the loading and merging/linking (but not interpret the result). The main interpreter pass would then simply run the interpreter on the input module and not bother with any loading or merging. Also, I think I'd implement that separation *first* and then rebase this change on top of it.

What do you think?

ftynse commented:
Approach B makes sense to me. Generally, anything that keeps the interpreter simpler and more focused is likely a good idea. I suppose it should be sufficient to just move the symbols from other files,normally, it is allowed to have multiple declarations as long as there is a single definition + handle name conflicts between _private_ symbols that may have repeated names in different modules. Then we can ran symbol-dce separately to remove unused symbols.

ftynse commented:
Implementation wise, we can consider having a "mergeModules" function that takes a list of modules and moves everything into the first one. It can then be exercised from the pass that loads transform library files, but is also reusable.

ingomueller-net commented:
> Implementation wise, we can consider having a "mergeModules" function that takes a list of modules and moves everything into the first one. It can then be exercised from the pass that loads transform library files, but is also reusable.

Yeah, good point! Being able to merge several modules is on my TODO list. That would also be required if we wanted to write a generic merger that could replace most of [SPIR-V's `ModuleCombiner`](https://github.com/llvm/llvm-project/blob/d5c8b23b/mlir/include/mlir/Dialect/SPIRV/Linking/ModuleCombiner.h#L70). But I'll leaver that to that after the remaining pending changes...

ingomueller-net commented:
Yes, but you can do (slightly) better by being careful when to (re)construct which symbol table(s)...

ingomueller-net committed:
Several minor and major clean-ups:

* Move all private functions of the CPP file into anonymous namespace.
* Remove test with second interpreter pass that reloads the library. I
  think that this shouldn't be possible.
* Factor out `renameToUnique`, `canMergeInto`, and `mergeInto` into
  proper functions.
* Use a single symbol table per input op and update it correctly
  whenever symbols or ops change.
* Make `other` arg an `OwningOpRef` and clone the arguments where
  necessary.
* Improve comments.

ingomueller-net commented:
I think I changed my mind about the question of whether or not loading should be part of the interpreter. At the very least, I see that it is absolutely consistent to think of the current pass as "inject dependencies and interpret result." In this case, it is clear that running that pass twice in a way that tries to inject the dependencies twice does not work. The PR thus currently removes that test. If we decide that injection should be separated from the interpreter or that injection should be done only if necessary (such that it is idempotent), I believe we can do that in some future PR.

Other than that, I have overhauled the initial commit and think that it is ready for review now.

ingomueller-net committed:
Minor fixes.

* Use `moveBefore` instead of `moveAfter` in order to work on empty
  targets as well.
* Do not verify the target after moving each op, since the last op may
  use symbols of ops that still have to be moved.

ftynse commented on mlir/lib/Dialect/Transform/Transforms/TransformInterpreterPassBase.cpp:
```diff
@@ -302,80 +304,268 @@ static void performOptionalDebugActions(
+  StringAttr oldName = op.getNameAttr();
+  StringAttr newName;
+  {
+    MLIRContext *context = op->getContext();
+    SmallString<64> prefix = oldName.getValue();
+    prefix.push_back('_');
+    while (true) {
+      newName = StringAttr::get(context, prefix + Twine(uniqueId++));
+      if (!symbolTable.lookup(newName) && !otherSymbolTable.lookup(newName)) {
+        break;
+      }
     }
```
I'm not very happy about this low-level renaming detail here. I suppose if this moves to `SymbolTable.h/cpp` as a utility that shares the renaming logic with the table, it will be okay. Only this part, not the the use replacement.

ftynse replied:
LLVM doesn't like ravioli commits.

ftynse replied:
Having a chain of commits, each with a small amount of code to avoid touching multiple subsystems. Here, it would be changing the SymbolTable in a separate commit. Some project follow this style, e.g., to have different code owners review them separately. LLVM/MLIR prefer having larger cross-cutting commits to see the overall motivation for changes.

No action is necessary (pull requests will be squashed into one commit anyway) so I resolved the discussion.

ftynse commented on mlir/lib/Dialect/Transform/Transforms/TransformInterpreterPassBase.cpp:
```diff
@@ -302,80 +304,268 @@ static void performOptionalDebugActions(
+  }
+
+  // Apply renaming.
+  LLVM_DEBUG(llvm::dbgs() << ", renaming to @" << newName.getValue() << "\n");
```
Nit: looks like the debug line starts with a comma for no reason.

ftynse commented on mlir/lib/Dialect/Transform/Transforms/TransformInterpreterPassBase.cpp:
```diff
@@ -302,80 +304,268 @@ static void performOptionalDebugActions(
+                             SymbolTable &otherSymbolTable, int &uniqueId) {
+  assert(symbolTable.lookup(op.getNameAttr()) == op &&
+         "symbol table does not contain op");
+  assert(otherSymbolTable.lookup(otherOp.getNameAttr()) == otherOp &&
+         "other symbol table does not contain other op");
```
Nit: can't we just use the symbol table to resolve the name and obtain `otherOp` instead of passing it in?

ftynse commented on mlir/lib/Dialect/Transform/Transforms/TransformInterpreterPassBase.cpp:
```diff
@@ -302,80 +304,268 @@ static void performOptionalDebugActions(
-      LLVM_DEBUG(llvm::dbgs() << "cannot compare types\n");
+/// Return whether `func1` can be merged into `func2`.
+bool canMergeInto(FunctionOpInterface func1, FunctionOpInterface func2) {
+  return func1.isExternal() && (func2.isPublic() || func2.isExternal());
```
This condition is unclear to me. Why can we merge an external function into another external function? Neither has a body... Why should the other function be public? Is this about merging _declarations_?

ftynse replied:
A more fitting name could be beneficial, but I can't come up with one immediately. `canMergeDeclarationInto`?

ingomueller-net replied:
Yeah, that could work. (As you probably also felt, it's not perfect: the name suggest that it *has* to be a declaration but it actually checks *if* it is a declaration.)

One argument in favor of keeping it as is is that this is a formulation that generalizes: In the SPIR-V module combiner, they "merge" global variables and constants, and those are "mergeable" if they define the same thing. This generalization could be formalized into a `MergeableSymbol` interface (which extends `Symbol`) with two methods `canMergeInto` and `mergeInto`. At least for these two examples (SPIR-V and TD), that concept seems to work...

qcolombet replied:
Kind of late to the party, but I agree that the meaning is difficult to grasp currently.
More specifically I believe the comment doesn't describe what mergeable means.

For me the confusing part is that merging implies that the result is bigger. So before reading the code I was wondering if we were doing some sort of function concatenation :).
\<digression>LLVM has an optimization pass called merge globals that puts global variables in the same global structure to allow faster symbol resolution and that's the kind of thing this function name brings to mind for me.\</digression>

Maybe `canCoalesceSymbols`?

ingomueller-net replied:
I am open to using a different verb or name. From what I understand from the dictionary, [unite](https://www.merriam-webster.com/dictionary/unite), [combine](https://www.merriam-webster.com/dictionary/combine), [merge](https://www.merriam-webster.com/dictionary/merge), and [coalesce](https://www.merriam-webster.com/dictionary/coalesce) all roughly mean "take several things and make a single one of them."

I guess that that is undoubtably what we do with the outer ops (and the result is indeed larger usually). I'd argue that it is *close* to what we do with the symbols as well but maybe not *exactly* the same, at least not what we do with functions currently. Another angle to look at what we do with symbols is that we "deduplicate" them. Maybe that can yield a name?

ftynse commented on mlir/lib/Dialect/Transform/Transforms/TransformInterpreterPassBase.cpp:
```diff
@@ -302,80 +304,268 @@ static void performOptionalDebugActions(
+  assert(func1->getParentOp() == func2->getParentOp() &&
+         "expected func1 and func2 to be in the same parent op");
+
+  MLIRContext *context = func1->getContext();
+  auto consumedName = StringAttr::get(
+      context, transform::TransformDialect::kArgConsumedAttrName);
+  auto readOnlyName = StringAttr::get(
+      context, transform::TransformDialect::kArgReadOnlyAttrName);
```
Nit: I wonder if we can construct this in the dialect initialization and have the dialect give us the attributes directly.

ftynse commented on mlir/lib/Dialect/Transform/Transforms/TransformInterpreterPassBase.cpp:
```diff
@@ -302,80 +304,268 @@ static void performOptionalDebugActions(
+/// one of the two is external, in which case the other op preserved (or any one
+/// of the two if both are external).
+static LogicalResult mergeSymbolsInto(Operation *target,
+                                      OwningOpRef<Operation *> other) {
```
Why do we need to pass an owning reference here?

ftynse replied:
I suspect that we may prefer cloning specific symbols from `other` in the longer run, as the entire module may be quite large and we don't necessarily need all of them. I'm fine with the current model if you an add the rationale above into the function documentation and something like `TODO: reconsider efficiency when we have enough users`.

ingomueller-net replied:
OK, done. The `TODO` says that cloning individual ops probably makes most sense if we combine it with some pruning technique of the symbols that should be merged.

ftynse commented on mlir/lib/Dialect/Transform/Transforms/TransformInterpreterPassBase.cpp:
```diff
@@ -302,80 +304,268 @@ static void performOptionalDebugActions(
+  // Rename private symbols in both ops in order to resolve conflicts that can
+  // be resolved that way.
+  LLVM_DEBUG(DBGS() << "renaming private symbols to resolve conflicts:\n");
+  for (auto [symbolTable, otherSymbolTable] : llvm::zip(
```
Nit: use `auto &&` for zippers.

ftynse commented on mlir/lib/Dialect/Transform/Transforms/TransformInterpreterPassBase.cpp:
```diff
@@ -302,80 +304,268 @@ static void performOptionalDebugActions(
+  SymbolTable targetSymbolTable(target);
+  SymbolTable otherSymbolTable(*other);
+
+  int uniqueId = 0;
```
I don't see the point of sharing the counter across different symbol names. Starting at 0 on each collision would produce more readable names.

ftynse commented on mlir/lib/Dialect/Transform/Transforms/TransformInterpreterPassBase.cpp:
```diff
@@ -302,80 +304,268 @@ static void performOptionalDebugActions(
+
+      LLVM_DEBUG(llvm::dbgs() << ", emitting error\n");
+      InFlightDiagnostic diag =
+          emitError(symbolOp->getLoc(),
```
Nit: prefer the `emitError(loc) << foo << bar` format. Twines are always tricky.

ftynse commented on mlir/lib/Dialect/Transform/Transforms/TransformInterpreterPassBase.cpp:
```diff
@@ -302,80 +304,268 @@ static void performOptionalDebugActions(
+  // be resolved that way.
+  LLVM_DEBUG(DBGS() << "renaming private symbols to resolve conflicts:\n");
+  for (auto [symbolTable, otherSymbolTable] : llvm::zip(
+           SmallVector<SymbolTable *>{&targetSymbolTable, &otherSymbolTable},
+           SmallVector<SymbolTable *>{&otherSymbolTable, &targetSymbolTable})) {
```
This place is the rare place to explicitly specify the number of stack elements in the vector.

ftynse commented on mlir/lib/Dialect/Transform/Transforms/TransformInterpreterPassBase.cpp:
```diff
@@ -302,80 +304,268 @@ static void performOptionalDebugActions(
-    builder.setInsertionPoint(&op);
-    builder.clone(*externalSymbol);
-    symbol->erase();
+  for (auto *op : SmallVector<Operation *>{target, *other}) {
+    if (failed(mlir::verify(op)))
+      return emitError(op->getLoc(),
+                       "failed to verify input op after renaming");
```
I don't think we should be running local verification here. Has it caught any errors?
Maybe we should consider having "merge" as a separate pass that runs before the interpreter so there is an opportunity for the verifier to run in between (and to be disabled for compiler performance-sensitive cases).

ftynse replied:
If this is short-lived, we can keep it. Basically, verifier already runs between passes and is intended for that, so I would prefer not having it run _inside_ the pass. We will end up duplicating the pass infrastructure...

I was discussing with @nicolasvasilache on a related subject, and it looks like having a separate pass for resolving symbols will be beneficial for generating reproducers. Not for this PR though!

ingomueller-net replied:
OK, I added a comment with a TODO.

ingomueller-net committed:
Add `SymbolTable::rename`.

ingomueller-net committed:
Use new functions in `SymbolTable`.

ingomueller-net committed:
Add `SymbolTable::renameToUnique`.

ingomueller-net committed:
Change tests to make sure external declaration is preserved.

ingomueller-net committed:
Address remaining issues raised by @ftynse.

ingomueller-net committed:
Factor out transform dialect attribute name construction.

ingomueller-net reviewed (commented):
Thanks a lot, @ftynse, for the detailed feedback! I just pushed a bunch of commits that should address your comments (modulo the questions I am submitting here).

ingomueller-net committed:
Remove library file name from the repro call.

Since the pass injects all definitions, providing the library again
isn't needed. Since that injection isn't idempotent, it actually isn't
even *possible* anymore, so this commits removes that argument.

ingomueller-net committed:
Update function argument and CLI descriptions.

ftynse commented on mlir/include/mlir/Dialect/Transform/IR/TransformDialect.td:
```diff
@@ -42,6 +42,14 @@ def Transform_Dialect : Dialect {
       constexpr const static ::llvm::StringLiteral
           kArgReadOnlyAttrName = "transform.readonly";
 
+      /// Above attribute names as `StringAttr`.
```
Nit: this comment is rendered, e.g., for autocompletion, where there is no "above" context. Prefer something more explanatory.

ftynse commented on mlir/lib/Dialect/Transform/Transforms/TransformInterpreterPassBase.cpp:
```diff
@@ -50,6 +50,8 @@ constexpr static llvm::StringLiteral kTransformDialectTagPayloadRootValue =
 constexpr static llvm::StringLiteral
     kTransformDialectTagTransformContainerValue = "transform_container";
 
+namespace {
```
MLIR has a preference for using static functions.

ftynse commented on mlir/lib/Dialect/Transform/Transforms/TransformInterpreterPassBase.cpp:
```diff
@@ -302,80 +294,233 @@ static void performOptionalDebugActions(
+/// has to be a declaration (aka has to be external) and `func2` either has to
+/// be a declaration as well, or it has to be public (otherwise, it wouldn't
+/// be visible by `func1`).
+bool canMergeInto(FunctionOpInterface func1, FunctionOpInterface func2) {
```
Nit: declare as static.

ingomueller-net replied:
OK, done.

ftynse commented on mlir/lib/IR/SymbolTable.cpp:
```diff
@@ -218,6 +218,79 @@ StringAttr SymbolTable::insert(Operation *symbol, Block::iterator insertPt) {
+  assert(lookup(from) == op && "current name does not resolve to op");
+  assert(lookup(to) == nullptr && "new name already exists");
+
+  if (failed(SymbolTable::replaceAllSymbolUses(op, to, getOp())))
```
I doubt this is somehow intended, especially given that functions seem to share the implementation via templates. Can you investigate more?

ingomueller-net replied:
OK, will do.

ingomueller-net replied:
For the record: fixed by #68320.

ftynse commented on mlir/lib/Dialect/Transform/Transforms/TransformInterpreterPassBase.cpp:
```diff
@@ -161,17 +163,9 @@ static llvm::raw_ostream &
                const Pass::Option<std::string> &debugTransformRootTag,
-               const Pass::Option<std::string> &transformLibraryFileName,
                StringRef binaryName) {
-  std::string transformLibraryOption = "";
-  if (!transformLibraryFileName.empty()) {
-    transformLibraryOption =
-        llvm::formatv(" {0}={1}", transformLibraryFileName.getArgStr(),
-                      transformLibraryFileName.getValue())
-            .str();
-  }
```
Why is this deleted?

ingomueller-net replied:
I wrote this in the commit message:

> Since the pass injects all definitions, providing the library again isn't needed. Since that injection isn't idempotent, it actually isn't even *possible* anymore, so this commits removes that argument.

I think no comment is needed in the code, here, right?

ingomueller-net committed:
Address comments from final review of @ftynse.

ingomueller-net reviewed (commented):
Thanks for the review! Now all comments should be addressed. I'll merge once CI has passed. Should some of the minor things still bother you/us, then I can still submit a fix-up.

ingomueller-net committed:
Add comment about verification in the middle of the pass.

qcolombet commented on mlir/lib/Dialect/Transform/Transforms/TransformInterpreterPassBase.cpp:
```diff
@@ -302,77 +294,236 @@ static void performOptionalDebugActions(
-    }
+/// Merge `func1` into `func2`. The two ops must be inside the same parent op
+/// and mergable according to `canMergeInto`. The function erases `func1` such
+/// that only `func2` exists when the function returns.
```
I'll add a comment about "metadata" merging rules (i.e., around consumed and readonly)

ingomueller-net replied:
You *will* or you *would*? I.e., who makes that commit?

qcolombet commented on mlir/lib/Dialect/Transform/Transforms/TransformInterpreterPassBase.cpp:
```diff
@@ -302,77 +294,236 @@ static void performOptionalDebugActions(
+/// the result. Public symbols may not collide, with the exception of
+/// instances of `SymbolOpInterface`, where collisions are allowed if at least
+/// one of the two is external, in which case the other op preserved (or any one
+/// of the two if both are external).
```
I think this comment would be easier to read if we break it down into several sentences (or bullet points?).

ingomueller-net replied:
Yep, sounds good. Let's have the person do it that adds the comment you suggested above.

qcolombet commented on mlir/test/Dialect/Transform/test-interpreter-external-symbol-def-invalid.mlir:
```diff
@@ -0,0 +1,14 @@
+// RUN: mlir-opt %s
```
Do we need a run line here?
If `lit` doesn't skip this file, maybe we could move all the "include" files in a sub-dir and tell lit to ignore these files.

Edit: I see that we have patterns of files like that already (e.g., `mlir/test/Dialect/Transform/test-interpreter-external-symbol-def.mlir`). I don't particularly like this kind of "fake" run line, but maybe we are actually testing something here?
@ftynse what do you think we should do?

ingomueller-net replied:
I copied this from other files and speculated that this is done to (1) make lit not complain and (2) maybe make a test fail if the syntax of this specific file gets out of date rather than relying a test of another file?

qcolombet reviewed (commented):
I told @ingomueller-net I was going to look at this, so here we are :).

Sorry for the delay. LGTM too, couple of comments inlined.

ingomueller-net reviewed (commented):
Thanks @qcolombet! The suggestions about the comments makes sense. Who does them?
